mavros/src/plugins/imu.cpp

```diff
@@ -0,0 +1,530 @@
+		}
+		else {
+		   c(0,0) = -1.0;
+		}
+	}
+
+	/**
+	 * @brief Fill and publish IMU data message.
+	 * @param[in] time_boot_ms		Message timestamp (not syncronized)
+	 * @param[in,out] orientation_enu	Orientation in the base_link ENU frame
+	 * @param[in,out] orientation_ned	Orientation in the aircraft NED frame
+	 * @param[in,out] gyro_enu		Angular velocity/rate in the base_link ENU frame
+	 * @param[in,out] gyro_ned		Angular velocity/rate in the aircraft NED frame
+	 */
+	void publish_imu_data(uint32_t time_boot_ms, Eigen::Quaterniond &orientation_enu,
+				Eigen::Quaterniond &orientation_ned, Eigen::Vector3d &gyro_enu, Eigen::Vector3d &gyro_ned)
+	{
+		auto imu_ned_msg = boost::make_shared<sensor_msgs::Imu>();
+		auto imu_enu_msg = boost::make_shared<sensor_msgs::Imu>();
+
+		// Fill message header
+		imu_enu_msg->header = m_uas->synchronized_header(frame_id, time_boot_ms);
+		imu_ned_msg->header = m_uas->synchronized_header("aircraft", time_boot_ms);
+
+		// Convert from Eigen::Quaternond to geometry_msgs::Quaternion
+		tf::quaternionEigenToMsg(orientation_enu, imu_enu_msg->orientation);
+		tf::quaternionEigenToMsg(orientation_ned, imu_ned_msg->orientation);
+
+		// Convert from Eigen::Vector3d to geometry_msgs::Vector3
+		tf::vectorEigenToMsg(gyro_enu, imu_enu_msg->angular_velocity);
+		tf::vectorEigenToMsg(gyro_ned, imu_ned_msg->angular_velocity);
+
+		// Eigen::Vector3d from HIGHRES_IMU or RAW_IMU, to geometry_msgs::Vector3
+		tf::vectorEigenToMsg(linear_accel_vec_enu, imu_enu_msg->linear_acceleration);
+		tf::vectorEigenToMsg(linear_accel_vec_ned, imu_ned_msg->linear_acceleration);
+
+		// Pass ENU msg covariances
+		imu_enu_msg->orientation_covariance = orientation_cov;
+		imu_enu_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_enu_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		// Pass NED msg covariances
+		imu_ned_msg->orientation_covariance = orientation_cov;
+		imu_ned_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_ned_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		/** Store attitude in base_link ENU
+		 *  @snippet src/plugins/imu.cpp store_enu
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_enu(imu_enu_msg);
+		// [store_enu]
+
+		/** Store attitude in aircraft NED
+		 *  @snippet src/plugins/imu.cpp store_ned
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_ned(imu_ned_msg);
+		// [store_ned]
+
+		/** Publish only base_link ENU message
+		 *  @snippet src/plugins/imu.cpp pub_enu
+		 */
+		// [pub_enu]
+		imu_pub.publish(imu_enu_msg);
+		// [pub_enu]
+	}
+
+	/**
+	 * @brief fill and publish IMU data_raw message; store linear acceleration for IMU data
+	 * @param[in] header		Message frame_id and timestamp
+	 * @param[in,out] gyro		Orientation in the base_link ENU frame
+	 * @param[in,out] accel_enu	Linear acceleration in the base_link ENU frame
+	 * @param[in,out] accel_ned	Linear acceleration in the aircraft NED frame
+	 */
```

Comment:
I do not see any assigns to any parameters. Why `in,out`?

Reply:
`[in,out]` is not explicit on the Doxygen docs. But I assume it is a parameter that is fed on the function, it is modified, and presents itself as an output (unless the `out` represents the output of `return` - is that it?).